runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -12358,6 +12358,936 @@ static TR::Register *inlineIntrinsicIndexOf(TR::Node *node, TR::CodeGenerator *c
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::bne, node, subStringMismatchLabel, cr6Reg);
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi2, node, subSearchOffsetReg, subSearchOffsetReg, 8);
+
+   /* If there are no more bytes, the arrays match each other. Jump to subStringMatchLabel to handle the substring match. */
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::subf_r, node, tempReg, subSearchOffsetReg, subStringLenReg);
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::beq, node, subStringMatchLabel, cr0Reg);
+
+   /* There are at least 2 bytes left. If there are less than 4 bytes left jump to the 2 byte load section. */
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::cmpli4, node, cr6Reg, tempReg, 4);
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::blt, node, skipLoad4Label, cr6Reg);
+
+   /* There are 4 to 7 bytes left at this point. */
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, skipLoad8Label);
+
+   /*
+    * Use 4 byte loads to search for a mismatch between the two arrays.
+    * There are 4 to 7 bytes left so this only needs to be done once.
+    */
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::lwzx, node, tempReg, subSearchOffsetReg, currentAddressReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::lwzx, node, temp2Reg, subSearchOffsetReg, subStringAddressReg);
+   /* If a mismatch is found, jump to subStringMismatchLabel to indicate that the substring was not found. */
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::cmp4, node, cr6Reg, tempReg, temp2Reg);
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::bne, node, subStringMismatchLabel, cr6Reg);
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi2, node, subSearchOffsetReg, subSearchOffsetReg, 4);
+
+   /* If there are no more bytes, the arrays match each other. Jump to subStringMatchLabel to handle the substring match. */
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::subf_r, node, tempReg, subSearchOffsetReg, subStringLenReg);
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::beq, node, subStringMatchLabel, cr0Reg);
+
+   /* There are exactly 2 bytes left at this point. */
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, skipLoad4Label);
+
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::lhzx, node, tempReg, subSearchOffsetReg, currentAddressReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::lhzx, node, temp2Reg, subSearchOffsetReg, subStringAddressReg);
+   /* If a match is found, jump to subStringMatchLabel to handle the substring match. */
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::cmp4, node, cr6Reg, tempReg, temp2Reg);
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::beq, node, subStringMatchLabel, cr6Reg);
+   /* A fallthrough means the loaded values did not match. Continue on to handle the case where the substring was not found. */
```

Comment:
i assumed the case where main string is shorter than sub-string is handled/guarded outside this routine.

Reply:
That is correct. It is handled in two places:

https://github.com/eclipse-openj9/openj9/blob/75f0b5ee0c00bfb4dde15664ba27beefa0d089b6/runtime/compiler/p/codegen/J9TreeEvaluator.cpp#L12646-L12649

https://github.com/eclipse-openj9/openj9/blob/75f0b5ee0c00bfb4dde15664ba27beefa0d089b6/runtime/compiler/p/codegen/J9TreeEvaluator.cpp#L12720-L12737

After the second section `endAddressReg` contains the address where the sub string is now too long to be found in the rest of the main string. As a result, it won't go to the code generated by `performArrayCmp` since a match is impossible.